File: packages/sim-core/src/observe.ts
import type { AgentObservation, AgentState, PolicyVector, WorldState } from "./types.js";
import { effectiveStringencyFromState } from "./regulatory.js";

/**
 * Build the serializable observation passed to policies (heuristic, QRE, LLM).
 * Includes wealth, knowledge, labor, reputation, neighbors, policy snapshot, memory, last profit, offering quality, and pending R&D count.
 */
export function buildObservation(
  agent: AgentState,
  world: WorldState,
): AgentObservation {
  const neighbors: { id: string; weight: number }[] = [];
  for (const e of world.edges) {
    if (e.a === agent.id) neighbors.push({ id: e.b, weight: e.weight });
    else if (e.b === agent.id) neighbors.push({ id: e.a, weight: e.weight });
  }
  const reg = world.config.regulatory;
  const erosion = reg.bribe.corruptionErodesStringency;
  const eff = reg.enabled
    ? effectiveStringencyFromState(
        world.regulatory.stringency,
        world.regulatory.corruption,
        erosion,
      )
    : 0;
  const sp = world.config.spawn;
  const pop = world.agents.length;
  const need = sp.parentCostWealth + sp.minParentWealthFloor;
  return {
    selfId: agent.id,
    type: agent.type,
    civicRole: agent.civicRole,
    publicServantFireable: agent.publicServantFireable,
    tick: world.tick,
    wealth: agent.wealth,
    knowledge: agent.knowledge,
    labor: agent.labor,
    patentCount: agent.patentExpiresAt.length,
    reputation: agent.reputation,
    neighbors,
    globalPool: world.globalPool,
    marketSize: world.marketSize,
    policy: world.config.policy as PolicyVector,
    regulatory: {
      enabled: reg.enabled,
      ruleMode: reg.ruleMode,
      stringency: world.regulatory.stringency,
      corruption: world.regulatory.corruption,
      effectiveStringency: eff,
      bribeEnabled: reg.bribe.enabled,
    },
    memory: [...agent.memory],
    lastProfit: agent.lastProfit,
    lastOfferingQuality: agent.lastOfferingQuality,
    pendingInnovationCount: agent.innovationPipeline.length,
    population: pop,
    spawn: {
      enabled: sp.enabled,
      maxAgents: sp.maxAgents,
      atCap: pop >= sp.maxAgents,
      canAffordSpawn: sp.enabled && pop < sp.maxAgents && agent.wealth >= need,
    },
  };
}

File: packages/ip-sim-engine/src/sim/observe.rs
//! Matches `packages/sim-core/src/observe.ts` (fields used by heuristic).

use crate::sim::regulatory::effective_stringency_from_state;
use crate::sim::types::{Agent, WorldState};

pub struct Observation {
    pub wealth: f64,
    pub patent_count: usize,
    pub reputation: f64,
    pub neighbors_len: usize,
    pub open_science_subsidy: f64,
    pub patent_regime_none: bool,
    pub enforcement_intensity: f64,
    pub spawn_enabled: bool,
    pub spawn_max: u32,
    pub spawn_at_cap: bool,
    pub spawn_can_afford: bool,
    pub regulatory_enabled: bool,
    pub bribe_enabled: bool,
}

pub fn build_observation(agent: &Agent, world: &WorldState) -> Observation {
    let mut neighbors_len = 0usize;
    for e in &world.edges {
        if e.a == agent.id || e.b == agent.id {
            neighbors_len += 1;
        }
    }
    let reg = &world.config.regulatory;
    let erosion = reg.bribe.corruption_erodes_stringency;
    let eff = if reg.enabled {
        effective_stringency_from_state(
            world.regulatory.stringency,
            world.regulatory.corruption,
            erosion,
        )
    } else {
        0.0
    };
    let _ = eff; // same as TS effectiveStringency in observation
    let sp = &world.config.spawn;
    let pop = world.agents.len() as u32;
    let need = sp.parent_cost_wealth + sp.min_parent_wealth_floor;
    Observation {
        wealth: agent.wealth,
        patent_count: agent.patent_expires_at.len(),
        reputation: agent.reputation,
        neighbors_len,
        open_science_subsidy: world.config.policy.open_science_subsidy,
        patent_regime_none: matches!(world.config.policy.patent_regime, super::types::PatentRegime::None),
        enforcement_intensity: world.config.policy.enforcement_intensity,
        spawn_enabled: sp.enabled,
        spawn_max: sp.max_agents,
        spawn_at_cap: pop >= sp.max_agents,
        spawn_can_afford: sp.enabled && pop < sp.max_agents && agent.wealth >= need,
        regulatory_enabled: reg.enabled,
        bribe_enabled: reg.bribe.enabled,
    }
}
